test: disable coverage in watch mode or when NO_COVERAGE env is present

shlomiassaf opened the pull request:
NO_COVERAGE is provide for IDE integration

This PR enables proper breaking point in TS code.

Fixes #1392

Approach:
All configurations files (karma/webpack.test) are configured without coverage, if coverage support is needed it will be added into the config's.
This is common sense, there's no point in setting something then removing it which is also not clear and readable.

Coverage is opt-out, i.e. by default you will get coverage reports... the developer needs to opt out of coverage explicitly.

There are 2 ways to opt out:
  - Using `--watch-mode`. watch mode implies development session so no coverage there. This is automatically set as part of the npm scripts in this project.

  - Using `NO_COVERAGE` environment variable, when set the coverage will not run. This is a fallback that is useful for IDE integration where karama is executed and controlled by the IDE.

![image](https://cloud.githubusercontent.com/assets/5377501/22733664/72be0e8a-edfb-11e6-9475-efdaa3ca91d7.png)

shlomiassaf commented:
After posting this I'v noticed that @yjaaidi has posted PR #1079

It appears to be roughly the same solution....

@yjaaidi my apologies...

I would prefer going with this one since it does not involve adding an extra package.

Please discuss.

gviligvili commented:
What about ? 
```
....


/**
 * Webpack Constants
 */
const ENV = process.env.ENV = process.env.NODE_ENV = 'test';

const autowatch = process.env.npm_lifecycle_script.indexOf('--auto-watch') !== -1;

....

/**
 * Webpack configuration
 *
 * See: http://webpack.github.io/docs/configuration.html#cli
 */
module.exports = function (options) {
      let config = {
              ...
      };

      // skip coverage in watch mode
      if (!autowatch) {
            config.module.rules.push(/**
                   /**
                   * Instruments JS files with Istanbul for subsequent code coverage reporting.
                   * Instrument only testing sources.
                   *
                   * See: https://github.com/deepsweet/istanbul-instrumenter-loader
                   */
                  {
                        enforce: 'post',
                        test: /\.(js|ts)$/,
                        loader: 'istanbul-instrumenter-loader',
                        include: helpers.root('src'),
                        exclude: [
                              /\.(e2e|spec)\.ts$/,
                              /node_modules/
                        ]
                  }
            );
      }

      return config;
};

```

This is how I disabled the coverage in watch mode,
Sorry if it's not right, just trying help.

shlomiassaf commented:
@gviligvili 

It's about the same, I just added support for an environment variable for IDEs.

I also removed some karma plugins, even though just removing istanbul does the trick there is still coverage plugins running which has no point...

MangeshPimpalkar commented:
Thanks it works great.

gdi2290 commented:
@shlomiassaf any updates on this?
